### PR TITLE
add filters to /job and /program

### LINKED ISSUE #1456: Gateway: implement sets of filter to return function by specific criteria
### What is the expected behavior?

We want to have `jobs` and `programs` endpoints to have additional filtering parameter (https://<host>/v1/jobs?filter=<filter_name>).

Initially we need only 2 filters (names might change):
- `providers`
  - for `/programs?filter=providers` will return function uploaded by providers only
  - for `/jobs?filter=providers` will return only user jobs associated with providers only
- `user`
  - for `/programs?filter=user` will return function uploaded by this user only
  - for `/jobs?filter=user` will return only user jobs associated with only user programs

2 filters above are mutually exclusive.

### PR DESCRIPTION
**Pull request by akihikokuroda**
### Summary
fix #1456

### CODE REVIEW

**Comment by IceKhan13** [comment]
Looks good! I'll try to run test it Monday for final approval :) Thanks Aki!

**Review by Tansito — COMMENTED** [review]
I see the logic fine @akihikokuroda but I would like to propose a change in the logic as I was mentioning here: https://github.com/Qiskit/qiskit-serverless/issues/1456#issuecomment-2302236300

Instead to use `catalog` & `serverless` I would use `user` & `provider` as Iskandar proposed. The reason behind this is that If other client would like to retrieve `user` or `partner` would need to use `catalog` or `serverless` what probably it has not much sense.

Apart from that I see everything ok 👍

File: gateway/api/views.py
Comment (Tansito) [inline gateway/api/views.py:342]: Just for curiosity what it does the `~` in front of the `Q`?

Comment (akihikokuroda) [reply gateway/api/views.py:342]: It's `NOT`.

**Comment by akihikokuroda** [comment]
@Tansito The filter values are used (sent) by the current qiskit_ibm_catalog side.   So they must be these values until the catalog client is changed. They are not visible from the users so we can create a refactor PR later and fix both client and gateway at the same time.

**Comment by Tansito** [comment]
Ooooh, ok, understood I didn't know that the catalog was already with those values.

**Review by IceKhan13 — APPROVED** [review]

**Comment by Tansito** [comment]
Don't wait for my approval to merge it @akihikokuroda , what I saw I saw it ok and if one of us approves it it's more than enough 👍

**Pull request merged**
